v2/pkg/engine/plan/datasource_configuration.go

```diff
@@ -33,83 +34,159 @@ type DataSourceConfiguration struct {
-	if d.hash != 0 {
-		return d.hash
-	}
-	d.hash = DSHash(xxhash.Sum64(d.Custom))
-	return d.hash
+type DirectivesConfigurations interface {
+	DirectiveConfigurations() *DirectiveConfigurations
 }
 
-type DataSourcePlannerConfiguration struct {
-	RequiredFields FederationFieldConfigurations
-	ProvidedFields *NodeSuggestions
-	ParentPath     string
-	PathType       PlannerPathType
-	IsNested       bool
+type NodesAccess interface {
+	ListRootNodes() TypeFields
+	ListChildNodes() TypeFields
 }
 
-type PlannerPathType int
-
-const (
-	PlannerPathObject PlannerPathType = iota
-	PlannerPathArrayItem
-	PlannerPathNestedInArray
-)
+type NodesInfo interface {
+	HasRootNode(typeName, fieldName string) bool
+	HasRootNodeWithTypename(typeName string) bool
+	HasChildNode(typeName, fieldName string) bool
+	HasChildNodeWithTypename(typeName string) bool
+}
 
-func (c *DataSourcePlannerConfiguration) HasRequiredFields() bool {
-	return len(c.RequiredFields) > 0
+func (d *DataSourceMetadata) DirectiveConfigurations() *DirectiveConfigurations {
+	return d.Directives
 }
 
-func (d *DataSourceConfiguration) HasRootNode(typeName, fieldName string) bool {
+func (d *DataSourceMetadata) HasRootNode(typeName, fieldName string) bool {
 	return d.RootNodes.HasNode(typeName, fieldName)
 }
 
-func (d *DataSourceConfiguration) HasRootNodeWithTypename(typeName string) bool {
+func (d *DataSourceMetadata) HasRootNodeWithTypename(typeName string) bool {
 	return d.RootNodes.HasNodeWithTypename(typeName)
 }
 
-func (d *DataSourceConfiguration) HasChildNode(typeName, fieldName string) bool {
+func (d *DataSourceMetadata) HasChildNode(typeName, fieldName string) bool {
 	return d.ChildNodes.HasNode(typeName, fieldName)
 }
 
-func (d *DataSourceConfiguration) HasChildNodeWithTypename(typeName string) bool {
+func (d *DataSourceMetadata) HasChildNodeWithTypename(typeName string) bool {
 	return d.ChildNodes.HasNodeWithTypename(typeName)
 }
 
-func (d *DataSourceConfiguration) HasKeyRequirement(typeName, requiresFields string) bool {
-	return d.FederationMetaData.Keys.HasSelectionSet(typeName, "", requiresFields)
+func (d *DataSourceMetadata) ListRootNodes() TypeFields {
+	return d.RootNodes
+}
+
+func (d *DataSourceMetadata) ListChildNodes() TypeFields {
+	return d.ChildNodes
+
+}
+
+// dataSourceConfiguration is the configuration for a DataSource
+type dataSourceConfiguration[T any] struct {
+	*DataSourceMetadata                   // DataSourceMetadata is the information about root and child nodes and federation metadata if applicable
+	ID                  string            // ID is a unique identifier for the DataSource
+	Factory             PlannerFactory[T] // Factory is the factory for the creation of the concrete DataSourcePlanner
+	Custom              json.RawMessage
+	NewCustom           T // NewCustom is the datasource specific configuration
+
+	hash DSHash // hash is a unique hash for the dataSourceConfiguration used to match datasources
```

Comment:
Please ensure that the hash is created during the configuration stage, not during the planning stage.
Once this is done, we can remove the pre-generation of DSHash at Router startup.